components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1218,6 +1218,20 @@ where
         exec_result
     }
 
+    fn need_pre_exec(&self, req: &RaftCmdRequest) -> bool {
+        if req.has_admin_request() {
```

Comment:
Why? I think all commands need to exec `pre_exec`.

Reply:
Yes, `pre_exec` is designed for all commands. If one want to pre_exec some kind of commands, he will register those in `need_pre_exec`.
The `need_pre_exec` is like a highlight for developers and reviewers. If a command like `BatchSplit` needs to be `pre_exec`, and thus be potentially skipped execution, we'd better think out why.

Reply:
Maybe I can pre_exec for every command, and skip when need_pre_exec is true.

Reply:
No, this is just a trap for observer authors.

Reply:
Then we just remove `need_pre_exec`?